### ISSUE TITLE
Symbol instantiation causes panic/UB

### ISSUE THREAD

**Issue by Ten0** [Robbepop/string-interner#19]
Following code fails to run on latest version:
```rust
use string_interner::{symbol::SymbolU16, Symbol};
fn main() {
	let s = SymbolU16::try_from_usize(std::u16::MAX as usize);
	dbg!(s);
}
```
It gives the following error:
```
thread 'main' panicked at 'attempt to add with overflow', .cargo/registry/src/github.com-1ecc6299db9ec823/string-interner-0.10.0/src/symbol.rs:64:77
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1069
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1504
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:511
  11: rust_begin_unwind
             at src/libstd/panicking.rs:419
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:111
  13: core::panicking::panic
             at src/libcore/panicking.rs:54
  14: <string_interner::symbol::SymbolU16 as string_interner::symbol::Symbol>::try_from_usize
             at .cargo/registry/src/github.com-1ecc6299db9ec823/string-interner-0.10.0/src/symbol.rs:64
```

That is because
https://github.com/Robbepop/string-interner/blob/85b1829b55496308bdc8bc100fe3eb2daf73e0d6/src/symbol.rs#L60-L68
is invalid in the `u16` case. (And I would guess it's Undefined Behaviour in release mode, as it calls `new_unchecked` on a non-zero type with a zero value).

I beleive the proper way to implement this, without `unsafe`, with the same number of instructions, and simpler, would be:
```rust
            fn try_from_usize(index: usize) -> Option<Self> {
                 <$non_zero>::new((index as $base_ty).wrapping_add(1))
            }
```

Also, `usize::MAX` only exists in very recent versions of Rust, so this would increase compatibility (otherwise triggers the following error:)
```
    Updating crates.io index
   Compiling string-interner v0.10.0
error[E0599]: no associated item named `MAX` found for type `usize` in the current scope
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/string-interner-0.10.0/src/symbol.rs:62:35
   |
62 |                   if index < usize::MAX {
   |                                     ^^^ associated item not found in `usize`
...
77 | / gen_symbol_for!(
78 | |     /// Symbol that is 16-bit in size.
79 | |     ///
80 | |     /// Is space-optimized for used in `Option`.
81 | |     struct SymbolU16(NonZeroU16; u16);
82 | | );
   | |__- in this macro invocation
   |
help: you are looking for the module in `std`, not the primitive type
   |
62 |                 if index < std::usize::MAX {
   |                            ^^^^^^^^^^^^^^^
error[E0599]: no associated item named `MAX` found for type `usize` in the current scope
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/string-interner-0.10.0/src/symbol.rs:62:35
   |
62 |                   if index < usize::MAX {
   |                                     ^^^ associated item not found in `usize`
...
83 | / gen_symbol_for!(
84 | |     /// Symbol that is 32-bit in size.
85 | |     ///
86 | |     /// Is space-optimized for used in `Option`.
87 | |     struct SymbolU32(NonZeroU32; u32);
88 | | );
   | |__- in this macro invocation
   |
help: you are looking for the module in `std`, not the primitive type
   |
62 |                 if index < std::usize::MAX {
   |                            ^^^^^^^^^^^^^^^

error[E0599]: no associated item named `MAX` found for type `usize` in the current scope
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/string-interner-0.10.0/src/symbol.rs:62:35
   |
62 |                   if index < usize::MAX {
   |                                     ^^^ associated item not found in `usize`
...
89 | / gen_symbol_for!(
90 | |     /// Symbol that is the same size as a pointer (`usize`).
91 | |     ///
92 | |     /// Is space-optimized for used in `Option`.
93 | |     struct SymbolUsize(NonZeroUsize; usize);
94 | | );
   | |__- in this macro invocation
   |
help: you are looking for the module in `std`, not the primitive type
   |
62 |                 if index < std::usize::MAX {
   |                            ^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors
```

**Comment by Robbepop** [comment]
Good find! Thanks for the bug report.
https://github.com/Robbepop/string-interner/blob/master/src/symbol.rs#L62
In this line the `usize` should actually be `$base_ty`.
Happy to merge a PR if you want to have the honor to fix it yourself. ;)

**Comment by Ten0** [comment]
> Happy to merge a PR if you want to have the honor to fix it yourself. ;)

Sure ! :)
I will try to open it tomorrow, otherwise the next day.

> In this line the `usize` should actually be `$base_ty`.

That would be one way to fix it, though I beleive it's much better to avoid `unsafe` usage here since it has no cost (may even be faster, not sure whether LLVM is able to optimize `if x == max {0} else {x+1}` into `x+1`, which is the form we directly get at with the implementation I'm proposing).

**Comment by Robbepop** [comment]
I just tested the 3 different versions of the same implementation:

1. Old one that is using `unsafe`:

```rust
impl Symbol for $name {
    fn try_from_usize(index: usize) -> Option<Self> {
        if index < <$base_ty>::MAX as usize {
            return Some(Self {
                value: unsafe { <$non_zero>::new_unchecked(index as $base_ty + 1) },
            })
        }
        None
    }
}
```

2. Your proposal:

```rust
impl Symbol for $name {
    fn try_from_usize(index: usize) -> Option<Self> {
        let shifted: $base_ty = if index < <$base_ty>::MAX as usize {
            index as $base_ty + 1
        } else {
            0
        };
        <$non_zero>::new(shifted).map(|non_zero| Self { value: non_zero })
    }
}
```

3. Branch-free implementation:

```rust
impl Symbol for $name {
    fn try_from_usize(index: usize) -> Option<Self> {
        <$non_zero>::new(
            (index < <$base_ty>::MAX as usize) as $base_ty * (index as $base_ty + 1)
        ).map(|non_zero| Self { value: non_zero })
    }
}
```

While in debug mode the branchless version is by far the fastest with optimizations all the above versions get compiled into the same machine code. So we should pick whatever version is the most readable.

**Comment by Ten0** [comment]
I believe we had a misunderstanding on my proposal : my proposal was an as branch-free as possible implementation one liner, in my first post, I was comparing to the current in my previous post.
The implementation I would recommend is:
```rust
            fn try_from_usize(index: usize) -> Option<Self> {
                 <$non_zero>::new((index as $base_ty).wrapping_add(1))
            }
```

**Comment by Robbepop** [comment]
> I believe we had a misunderstanding on my proposal : my proposal was an as branch-free as possible implementation one liner, in my first post, I was comparing to the current in my previous post.
> The implementation I would recommend is:
> 
> ```rust
>             fn try_from_usize(index: usize) -> Option<Self> {
>                  <$non_zero>::new((index as $base_ty).wrapping_add(1))
>             }
> ```

Ah yeah that's even a lot simpler! :sweat_smile: 
I have fixed the bug but will leave the code ugly so that you can file your PR with your proposal. :)

**Comment by Robbepop** [comment]
Fixed in [v0.10.1](crates.io/crates/string-interner).

**Comment by Robbepop** [comment]
I will close this issue now to not cause confusions but please remember your PR. ;)
